Fix: unique contraint in auto index creation

pedrocarlo opened the pull request:
Closes #1457 .

```sql
limbo> CREATE table t2 (x INTEGER PRIMARY KEY, y INTEGER UNIQUE);
limbo> SELECT * FROM sqlite_schema;
┌───────┬───────────────────────┬──────────┬──────────┬───────────────────────────────────────────────────────────┐
│ type  │ name                  │ tbl_name │ rootpage │ sql                                                       │
├───────┼───────────────────────┼──────────┼──────────┼───────────────────────────────────────────────────────────┤
│ table │ t2                    │ t2       │        2 │ CREATE TABLE t2 (x INTEGER PRIMARY KEY, y INTEGER UNIQUE) │
├───────┼───────────────────────┼──────────┼──────────┼───────────────────────────────────────────────────────────┤
│ index │ sqlite_autoindex_t2_1 │ t2       │        3 │                                                           │
└───────┴───────────────────────┴──────────┴──────────┴───────────────────────────────────────────────────────────┘
```

jussisaurio commented:
This goes into an infinite loop in your branch, known existing issue or bug in this?

```sql
limbo> CREATE table t2 (x INTEGER PRIMARY KEY, y INTEGER UNIQUE);
limbo> SELECT * FROM sqlite_schema;
┌───────┬───────────────────────┬──────────┬──────────┬───────────────────────────────────────────────────────────┐
│ type  │ name                  │ tbl_name │ rootpage │ sql                                                       │
├───────┼───────────────────────┼──────────┼──────────┼───────────────────────────────────────────────────────────┤
│ table │ t2                    │ t2       │        2 │ CREATE TABLE t2 (x INTEGER PRIMARY KEY, y INTEGER UNIQUE) │
├───────┼───────────────────────┼──────────┼──────────┼───────────────────────────────────────────────────────────┤
│ index │ sqlite_autoindex_t2_1 │ t2       │        3 │                                                           │
└───────┴───────────────────────┴──────────┴──────────┴───────────────────────────────────────────────────────────┘
limbo> insert into t2 (y) values (1),(1);
```

This works incorrectly:

```sql
limbo> CREATE table t2 (x INTEGER PRIMARY KEY, y INTEGER UNIQUE);
limbo> insert into t2 (y) values (1);
limbo> insert into t2 (y) values (6);
  × Runtime error: UNIQUE constraint failed: duplicate key
```

pedrocarlo commented:
It works incorrectly due to a bug in Insert. We should always have an autoindex created when a table has an unique constraint. 

```sql
limbo> explain insert into t2 (y) values (1);
addr  opcode             p1    p2    p3    p4             p5  comment
----  -----------------  ----  ----  ----  -------------  --  -------
0     Init               0     14    0                    0   Start at 14
1     OpenWrite          0     2     0                    0
2     Integer            1     3     0                    0   r[3]=1
3     OpenWrite          1     3     0                    0
4     NewRowid           0     1     0                    0   r[1]=rowid
5     Copy               2     5     0                    0   r[5]=r[2]
6     Copy               1     6     0                    0   r[6]=r[1]
7     MakeRecord         5     2     7                    0   r[7]=mkrec(r[5..6])
8     NoConflict         1     10    5     1              0   key=r[5]
9     Halt               1555  0     0     t2.x           0
10    IdxInsert          1     7     5                    0   key=r[7]
11    MakeRecord         2     2     4                    0   r[4]=mkrec(r[2..3])
12    Insert             0     4     1                    0
13    Halt               0     0     0                    0
14    Transaction        0     1     0                    0   write=true
15    Null               0     2     0                    0   r[2]=NULL
16    Goto               0     1     0                    0
```

pedrocarlo commented:
Investigating further I can see that when the index is created, its created only for the primary key column and not for the unique column. 
<img width="1069" alt="image" src="https://github.com/user-attachments/assets/9bdc2af6-3d09-4f87-bc33-3cd6fe6441d4" />

pedrocarlo commented:
Solved the problem. The issue was mainly that after I modified the check for index creation, the `automatic_from_primary_key_and_unique` function, was incorrectly constructing the index ignoring unique columns. Now I modified the logic to account for that.

jussisaurio commented on testing/insert.test:
```diff
@@ -161,3 +161,19 @@ do_execsql_test_on_specific_db {:memory:} strict-update-expression {
+|6}
+
+do_execsql_test_on_specific_db {:memory:} unique_insert_with_pkey {
+    CREATE TABLE t2 (x INTEGER PRIMARY KEY, y INTEGER UNIQUE);
```
What about cases like

```
CREATE TABLE t(x primary key, y unique);
```

This should result in two indexes being created - one for the PK, one for the unique column:

```
sqlite> select * from sqlite_schema;
table|t|t|2|CREATE TABLE t(x primary key, y unique)
index|sqlite_autoindex_t_1|t|3|
index|sqlite_autoindex_t_2|t|4|
```

jussisaurio replied:
Then there are cases like:

```sql
sqlite> create table t(a primary key, b unique, c,d, unique(c,d));
sqlite> select * from sqlite_schema;
table|t|t|2|CREATE TABLE t(a primary key, b unique, c,d, unique(c,d))
index|sqlite_autoindex_t_1|t|3|
index|sqlite_autoindex_t_2|t|4|
index|sqlite_autoindex_t_3|t|5|
```

pedrocarlo replied:
Right, I forgot it created indexes for each set of unique constraints. Sorry totally slipped my mind these cases. Right now, it's 3 am here, so I'm going to correct this tomorrow.

pedrocarlo commented:
@jussisaurio, ok I think now I got it sorted. This PR started with me thinking that this would be a simple fix, it involved me actually creating the correct amount of auto indices for Unique and Primary Key.

jussisaurio commented on core/schema.rs:
```diff
@@ -283,6 +285,7 @@ fn create_table(
     let mut primary_key_columns = vec![];
     let mut cols = vec![];
     let is_strict: bool;
+    let mut unique_sets: Option<Vec<Vec<(String, SortOrder)>>> = None;
```
would it be simpler if this was just a `Vec<Vec<(String, SortOrder)>>`? Empty Vec doesn't heap allocate

pedrocarlo replied:
Oh I thought it allocated. Alright let me change it

jussisaurio commented on core/schema.rs:
```diff
@@ -309,6 +312,31 @@ fn create_table(
+                            })
+                            .collect();
+                        if let Some(ref mut unique_sets) = unique_sets {
+                            unique_sets.push(unique_set);
```
one edge case: sqlite allows, but deduplicates, a multi-column unique index that is specified multiple times:

```sql
sqlite> create table t(a,b, unique(a,b), unique(a,b));
sqlite> select * from sqlite_schema;
table|t|t|2|CREATE TABLE t(a,b, unique(a,b), unique(a,b))
index|sqlite_autoindex_t_1|t|3|
```

pedrocarlo replied:
Alright let, me do that as well

pedrocarlo replied:
Ok, I think I managed to do it. The additional logic is a bit annoying and personally would much prefer this to be caught in the parser. Strange why sqlite does this. Maybe it is hard to catch this duplication in the parser?

jussisaurio commented on core/schema.rs:
```diff
@@ -764,44 +806,147 @@ impl Index {
+        // I wanted to just chain the iterator above but Rust type system get's messy with Iterators.
+        // It would not allow me chain them even by using a core::iter::empty()
+        // To circumvent this, I'm having to allocate a second Vec, and extend the other from it.
+        if table.get_rowid_alias_column().is_none() && !table.primary_key_columns.is_empty() {
```
would be good to extract this out as:

```rust
let has_primary_key_index = table.get_rowid_alias_column().is_none() && !table.primary_key_columns.is_empty();
if has_primary_key_index {
  ...
}

pedrocarlo commented on core/schema.rs:
```diff
@@ -309,6 +312,31 @@ fn create_table(
+                                let col_name = match column.expr {
+                                    Expr::Id(id) => normalize_ident(&id.0),
+                                    Expr::Literal(Literal::String(value)) => {
+                                        value.trim_matches('\'').to_owned()
```
Sorry was a left over from me copying the Primary Key pattern above it

pedrocarlo commented on core/schema.rs:
```diff
@@ -309,6 +312,31 @@ fn create_table(
                         }
+                    } else if let limbo_sqlite3_parser::ast::TableConstraint::Unique {
+                        columns,
+                        conflict_clause: _, // TODO: ignore conflict_cause for now
```
Done

jussisaurio commented on core/schema.rs:
```diff
@@ -274,6 +274,30 @@ impl PseudoTable {
+
+impl PartialEq for UniqueColumnProps {
+    fn eq(&self, other: &Self) -> bool {
+        self.column_name.eq(&other.column_name)
```
keep in mind that e.g. `unique(a desc, b)` is legal, so `SortOrder` should be part of these too

pedrocarlo replied:
Yeah, but they are deduped even if they have a `SortOrder`.

```sql
sqlite> create table t(a,b, unique(a,b), unique(a asc,b));
sqlite> SELECT * from sqlite_schema;
table|t|t|2|CREATE TABLE t(a,b, unique(a,b), unique(a asc,b))
index|sqlite_autoindex_t_1|t|3|
```

jussisaurio replied:
Ah cool, ok 👍

jussisaurio commented:
@pedrocarlo do we have a fix for this infinite loop in another PR?

```sql
limbo> CREATE table t2 (x INTEGER PRIMARY KEY, y INTEGER UNIQUE);
limbo> insert into t2 (y) values (1),(1); -- goes into infinite loop
```

pedrocarlo commented:
@jussisaurio just saw your comment. Good that you fixed it. I have to check later if we are emitting all of the same instructions as SQLite for insert multiple values.  But I think this is better dedicated to another PR.
